### PR TITLE
DAOS-7123 dfuse: Fix race condition in dfuse.

### PR DESCRIPTION
**Pull request by ashleypittman**
Correctly protect against simultanously connecting to and
forgetting a container by using hash tables to track pool
and container connections rather than a simple list.

What could happen was that a forget could be issued, causing
an inode to be closed and a container to be closed, but a
different thread could have already checked the pool list and
decided to re-use the pool.  Lack of ref counting on the list
entries meant this was unsafe, so the switch is to change to
a reference counted hash table which avoids the issue and is
generally safer.

### CODE REVIEW

File: src/client/dfuse/dfuse_cont.c [hunk 1]
```diff
@@ -27,123 +28,79 @@ dfuse_cont_open(fuse_req_t req, struct dfuse_inode_entry *parent,
...
 	D_ALLOC_PTR(ie);
 	if (!ie)
 		D_GOTO(close, rc = ENOMEM);
 
-	ie->ie_root = true;
-
 	DFUSE_TRA_UP(ie, parent, "inode");
 
-	rc = dfs_lookup(dfs->dfs_ns, "/", O_RDWR, &ie->ie_obj, NULL,
+	rc = dfs_lookup(dfc->dfs_ns, "/", O_RDWR, &ie->ie_obj, NULL,
```
Comment (daltonbohning) [inline src/client/dfuse/dfuse_cont.c:103]: Above, there is `d_hash_rec_find(&fs_handle->dpi_iet, ...)`. And if found, we copy in the inode entry. So this block is what happens if `!rlink`. But we're looking up root "/"? And we aren't inserting anything? I don't fully see the connection between the hash find above and this block of code below.

File: src/client/dfuse/dfuse_cont.c [hunk 2]
```diff
@@ -155,33 +112,27 @@ dfuse_cont_open(fuse_req_t req, struct dfuse_inode_entry *parent,
...
-	D_MUTEX_UNLOCK(&fs_handle->dpi_info->di_lock);
 	return;
-
 close:
-	daos_cont_close(dfs->dfs_coh, NULL);
 	D_FREE(ie);
-err_unlock:
-	D_MUTEX_UNLOCK(&fs_handle->dpi_info->di_lock);
+decref:
+	d_hash_rec_decref(&dfp->dfp_cont_table, &dfc->dfs_entry);
```
Comment (daltonbohning) [inline src/client/dfuse/dfuse_cont.c:126]: Is this safe to call when `if (rlink)` above is false? I.e. when we don't find an entry in the hash

File: src/client/dfuse/dfuse_pool.c [hunk 3]
```diff
@@ -35,100 +34,56 @@ dfuse_pool_lookup(fuse_req_t req, struct dfuse_inode_entry *parent,
...
-	dfs->dfs_dfp = dfp;
+	/* Drop the reference on the pool */
+	d_hash_rec_decref(&fs_handle->dpi_pool_table, &dfp->dfp_entry);
 
-	DFUSE_TRA_UP(dfp, parent->ie_dfs->dfs_dfp, "dfp");
+	rlink = d_hash_rec_find(&fs_handle->dpi_iet,
+				&dfc->dfs_ino,
+				sizeof(dfc->dfs_ino));
+	if (rlink) {
+		struct fuse_entry_param	entry = {0};
```
Comment (daltonbohning) [inline src/client/dfuse/dfuse_pool.c:68]: Similar to the container code, if we don't find the pool, should we be inserting it below?

**Review by daltonbohning — COMMENTED** [review]
I don't fully follow everything that's happening, but I think that's because I don't fully understand the internals of dfuse.

Are there or will there be tests in place to ensure the hash handles and references are accounted for and cleaned up as expected?

File: src/client/dfuse/dfuse_cont.c [hunk 4]
```diff
@@ -27,123 +28,79 @@ dfuse_cont_open(fuse_req_t req, struct dfuse_inode_entry *parent,
...
+
+		DFUSE_TRA_INFO(ie,
+			       "Reusing existing container entry without reconnect");
+
+		/* Update the stat information, but copy in the
+		 * inode value afterwards.
+		 */
+		rc = dfs_ostat(ie->ie_dfs->dfs_ns, ie->ie_obj,
+			       &entry.attr);
+		if (rc) {
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/dfuse_cont.c:81]: True, will fix on next push.

File: src/client/dfuse/dfuse_cont.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfuse/dfuse_cont.c:103]: Yes, this is for lookup where the result is the container.  The d_hash_rec_find() is looking up the inode for the container, if it's found (it's in cache) then it's update with the ostat and the function returns.  This particularly code-path is called the first time the container is opened only, so it does a lookup of / to find the root entry, later on this calls dfuse_reply_entry() which does a find_insert() of the inode where the insert will happen.

File: src/client/dfuse/dfuse_cont.c (hunk 2 above)
Comment (ashleypittman) [reply src/client/dfuse/dfuse_cont.c:126]: Yes, the rec_find() above is for looking up the inode, but this reference is for the container handle itself.

File: src/client/dfuse/dfuse_pool.c [hunk 5]
```diff
@@ -181,25 +136,15 @@ dfuse_pool_lookup(fuse_req_t req, struct dfuse_inode_entry *parent,
...
 
-	d_list_add(&dfp->dfp_list, &fs_handle->dpi_info->di_dfp_list);
-
-	dfs->dfs_ino = atomic_fetch_add_relaxed(&fs_handle->dpi_ino_next, 1);
-
-	ie->ie_stat.st_ino = dfs->dfs_ino;
-	dfs->dfs_ops = &dfuse_cont_ops;
+	ie->ie_stat.st_ino = dfc->dfs_ino;
 
 	dfuse_reply_entry(fs_handle, ie, NULL, true, req);
```
Comment (ashleypittman) [inline src/client/dfuse/dfuse_pool.c:141]: The insert is done in this function.

File: src/client/dfuse/dfuse_cont.c (hunk 1 above)
Comment (daltonbohning) [reply src/client/dfuse/dfuse_cont.c:103]: Ah, that makes sense!

**Comment by ashleypittman** [comment]
> Are there or will there be tests in place to ensure the hash handles and references are accounted for and cleaned up as expected?

No new tests for this PR no, the existing tests/code are failing very rarely but I haven't been able to trigger the kernel behaviour to write a reliable reproducer.

This PR has code in it (dfuse_cont_close_cb and dfuse_pool_close_cb) where it iterates the hash tables and logs any entries as errors during shutdown, so any lost references would result is warning messages and memory leaks, none of which remain.

File: src/client/dfuse/dfuse_cont.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfuse/dfuse_cont.c:103]: It depends on what it's looking up, but for containers/pools it does an initial find, then if that fails it allocates everything and does a find_insert(), for dfs objects they're cheaper and the inode hash table is bigger so it always allocates everything, does a single find_insert() and then either frees or re-uses as needed.

The _find for containers/pools is the common case and they're infrequently updated so it makes sense from a performance perspective, they're completely optional however and were the last bit of code I wrote for this PR so I could also check for leaks/cleanup in the find_insert() paths.  If _find doesn't get anything but find_insert() comes back with something then there are two threads doing lookups concurrently.

File: src/client/dfuse/dfuse_cont.c [hunk 6]
```diff
@@ -27,123 +28,79 @@ dfuse_cont_open(fuse_req_t req, struct dfuse_inode_entry *parent,
...
-			    DAOS_COO_RW, &dfs->dfs_coh, NULL, NULL);
-	if (rc == -DER_NONEXIST) {
-		DFUSE_TRA_INFO(dfs, "daos_cont_open() failed: (%d)", rc);
-		D_GOTO(err_unlock, rc = daos_der2errno(rc));
-	} else if (rc != -DER_SUCCESS) {
-		DFUSE_TRA_ERROR(dfs, "daos_cont_open() failed: (%d)", rc);
-		D_GOTO(err_unlock, rc = daos_der2errno(rc));
-	}
+	rc = dfuse_cont_open(fs_handle, dfp, &cont, &dfc);
+	if (rc)
```
Comment (daltonbohning) [inline src/client/dfuse/dfuse_cont.c:62]: If you push again...
```suggestion
	if (rc != 0)
```

**Review by daltonbohning — APPROVED** [review]
Assuming the current CI tests pass, LGTM.

**Review by mchaarawi — COMMENTED** [review]
it would be nice to add a test case that can potentially expose this race condition you are fixing

File: src/client/dfuse/dfuse_cont.c [hunk 7]
```diff
@@ -27,123 +28,79 @@ dfuse_cont_open(fuse_req_t req, struct dfuse_inode_entry *parent,
...
-	rc = daos_cont_open(dfp->dfp_poh, dfs->dfs_cont,
-			    DAOS_COO_RW, &dfs->dfs_coh, NULL, NULL);
-	if (rc == -DER_NONEXIST) {
-		DFUSE_TRA_INFO(dfs, "daos_cont_open() failed: (%d)", rc);
-		D_GOTO(err_unlock, rc = daos_der2errno(rc));
-	} else if (rc != -DER_SUCCESS) {
-		DFUSE_TRA_ERROR(dfs, "daos_cont_open() failed: (%d)", rc);
-		D_GOTO(err_unlock, rc = daos_der2errno(rc));
-	}
+	rc = dfuse_cont_open(fs_handle, dfp, &cont, &dfc);
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/dfuse_cont.c:61]: There's a comment by dfuse_cont_open() about this, dfc is both an in/out parameter, if it's valid when the call is made the the container and dfs handles are re-used, so there shouldn't be calls to dfs_cont_create() followed by cont_open/dfs_mount.  I realised after I'd written it that I could have made dfuse_cont_open() also take a "create" parameter to do the same thing which might have been better.

**Comment by ashleypittman** [comment]
> it would be nice to add a test case that can potentially expose this race condition you are fixing

We found it through testing but that I'm aware of I've only seen it fail twice, once on master and once on the release branch.

I did have a PR that simply repeated the specific test 100 times, and when I first pushed it there was a failure on iteration 17 but since a merge I've never got it to fail again.

**Pull request merged**